### PR TITLE
test(nemesis): mv range tombstones

### PR DESCRIPTION
**Pull request by juliayakovlev**
Add new nemesis will cover scenario is described in https://github.com/scylladb/scylla-enterprise/issues/3072.

When a base table of a materialized view is updated, the affected rows are also changed in the
materialized view. For DELETE statements, many rows can be affected, and so the view update
code splits the work into batches. This split was not performed correctly when range tombstones
were involved.
When the view_updating_consumer exceeds its buffer size limit, it flushes the mutation fragment
stream in the middle of a partition. But it doesn't take care to maintain range tombstones
properly while doing this, and if the buffer limit is exceeded in the middle of a range tombstone,
the mutation fragment stream will end with an unclosed range tombstone, which is illegal.

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I didn't leave commented-out/debugging code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

**Comment by juliayakovlev** [comment]
The issue is reproduced by this run: https://jenkins.scylladb.com/job/scylla-staging/job/yulia/job/yulia-longevity-large-partition-8h-test/72/

File: longevity_large_partition_test.py
Comment (fruch) [inline longevity_large_partition_test.py:22]: Why it was moved from configuration to here?

File: sdcm/nemesis.py
Comment (fruch) [inline sdcm/nemesis.py:3914]: We need a way to make this one work only on large partition cases

File: sdcm/nemesis.py
Comment (fruch) [inline sdcm/nemesis.py:3925]: Can we just select all test keyspaces ?

File: sdcm/nemesis.py
Comment (fruch) [inline sdcm/nemesis.py:5634]: Need to generate the nemesis list and the pipeline for this nemesis

File: longevity_large_partition_test.py
Comment (juliayakovlev) [reply longevity_large_partition_test.py:22]: Roy asked to add it to all large partition jobs

File: sdcm/nemesis.py
Comment (juliayakovlev) [reply sdcm/nemesis.py:3914]: I do not sure how this nemesis connected to the commit

File: sdcm/nemesis.py
Comment (juliayakovlev) [reply sdcm/nemesis.py:3925]: This is the goal here - destroy data in the table with large partitions and MV

File: sdcm/nemesis.py
Comment (juliayakovlev) [reply sdcm/nemesis.py:5634]: I will do it once when it will be ready and implementation and nemesis name will be approved

**Comment by juliayakovlev** [comment]
Test with Scylla master passed: https://jenkins.scylladb.com/job/scylla-staging/job/yulia/job/yulia-longevity-large-partition-8h-test/73/

**Comment by juliayakovlev** [comment]
Added changes according to @roydahan requests (described in task)

Test ran with 2022.2.6:
https://jenkins.scylladb.com/job/scylla-staging/job/yulia/job/yulia-longevity-large-partition-8h-test/78/

Screenshot from 2023-07-30 11-32-26

Issue https://github.com/scylladb/scylla-enterprise/issues/3072 reproduced

**Comment by juliayakovlev** [comment]
The test was run with latest master. All nemeses are passed

https://jenkins.scylladb.com/job/scylla-staging/job/yulia/job/yulia-longevity-large-partition-8h-test/79/

**Comment by juliayakovlev** [comment]
@fruch @roydahan 

Do we want to run the all jobs that I touched by their yaml?

File: test-cases/longevity/longevity-large-partition-8h.yaml
Comment (fruch) [inline test-cases/longevity/longevity-large-partition-8h.yaml:61]: Put those back

File: sdcm/nemesis.py
Comment (fruch) [inline sdcm/nemesis.py:1118]: Why are you leaving this comment?

**Review by fruch — CHANGES_REQUESTED** [review]
Clear the commented out code

And make sure all the testing changes are taken out of the yamls

**Comment by fruch** [comment]
> @fruch @roydahan 
> 
> Do we want to run the all jobs that I touched by their yaml?

We might need to validate they are working for a few hours, since this extra load might tip the use case off

File: sdcm/nemesis.py
Comment (juliayakovlev) [reply sdcm/nemesis.py]: Missed it, thanks

File: test-cases/longevity/longevity-large-partition-8h.yaml
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-8h.yaml]: Missed it, thanks

**Comment by roydahan** [comment]
Don't run all tests for now.

File: sdcm/nemesis.py
Comment (roydahan) [inline sdcm/nemesis.py:3915]: So eventually we do want to have this special nemesis?
I thought it's enough with the existing ones we have in case they happen in certain sequence.

File: test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml:1]: we shouldn't have this file at all.
Since we moved to i4i, this one is exactly as the other one.

File: test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml
Comment (roydahan) [inline test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml:33]: I'm not sure about that.
Need to think of it.

On one hand, it's good that we have data that we re-write after it was deleted.
OTHO, it will be the entire data eventually.

Maybe, we want to re-write only one part and to add more partitions on beyond the original range.

File: sdcm/nemesis.py
Comment (roydahan) [inline sdcm/nemesis.py:1025]: This part that handle the file name should be split into a separate commit.
We need this for all "DestroyData" nemesis.

Comment (roydahan) [reply sdcm/nemesis.py:1025]: (and also the parts below) but without the new nemesis.

Comment (juliayakovlev) [reply sdcm/nemesis.py:1025]: If so it's better to create a separate PR

File: test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml:1]: there are 7 jobs with this yaml
Screenshot from 2023-08-01 18-47-06

File: test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml:33]: > I'm not sure about that. Need to think of it.
> 
> On one hand, it's good that we have data that we re-write after it was deleted. OTHO, it will be the entire data eventually.
> 
> Maybe, we want to re-write only one part and to add more partitions on beyond the original range.

There are load already defined in the test that add different types of partitions (not same as is inserted while prepare stage)

File: sdcm/nemesis.py
Comment (fruch) [reply sdcm/nemesis.py:1025]: let's do it in a separate PR

File: sdcm/nemesis.py
Comment (fruch) [reply sdcm/nemesis.py:3915]: I think it's o.k. to leave this nemesis for a known issue, but at least to add the issue to the docstring

File: sdcm/nemesis.py
Comment (juliayakovlev) [reply sdcm/nemesis.py]: done

File: test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml]: test removed

**Comment by juliayakovlev** [comment]
longevity-large-partition-8h-test test passed:
https://jenkins.scylladb.com/job/scylla-staging/job/yulia/job/yulia-longevity-large-partition-8h-test/85/

File: test-cases/longevity/longevity-large-partition-3h.yaml
Comment (fruch) [inline test-cases/longevity/longevity-large-partition-3h.yaml:35]: @juliayakovlev why do we need the comment ? (last time it was copied wrongly to indexes, and failed, cause indexes doesn't support comments)

**Review by fruch — APPROVED** [review]
LGTM

File: test-cases/longevity/longevity-large-partition-3h.yaml
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-3h.yaml:35]: It is not something must. But it is allowed for materialized view, not like for index

**Comment by roydahan** [comment]
I'l re-review later today.

File: data_dir/nemesis.yml
Comment (roydahan) [inline data_dir/nemesis.yml:127]: Let's remove this new nemesis.
I think that the dtest of it is good enough to cover it and all the other changes here that increase the probability to hit these kind of issues randomly are good enough.

File: test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml
Comment (juliayakovlev) [reply test-cases/longevity/longevity-large-partition-200k_pks-2days-i4i.yaml]: Changed it to re-write 2/3 of available partitions. 1/3 of partitions won't be re-written and will remain as tombstones after deletion

File: data_dir/nemesis.yml
Comment (juliayakovlev) [reply data_dir/nemesis.yml]: Removed

**Comment by juliayakovlev** [comment]
@roydahan can you review, please?

**Comment by fruch** [comment]
@roydahan waiting for you final review on this one

**Comment by roydahan** [comment]
LGTM

**Review by fruch — APPROVED** [review]
LGTM

**Comment by yarongilor** [comment]
@juliayakovlev , i encountered a related failure in a staging job:
```
2023-08-31 09:15:37.584: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=53b45899-6b93-4012-9d33-e42316b7818d, source=LongevityTest.test_custom_time (longevity_test.LongevityTest)() message=Traceback (most recent call last):
File "/home/ubuntu/scylla-cluster-tests/longevity_test.py", line 106, in test_custom_time
self.run_prepare_write_cmd()
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 209, in run_prepare_write_cmd
self.run_post_prepare_cql_cmds()
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 152, in run_post_prepare_cql_cmds
self._run_cql_commands(post_prepare_cql_cmds)
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 143, in _run_cql_commands
session.execute(cmd)
File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/common.py", line 1751, in execute_verbose
return execute_orig(*args, **kwargs)
File "cassandra/cluster.py", line 2699, in cassandra.cluster.Session.execute
File "cassandra/cluster.py", line 5018, in cassandra.cluster.ResponseFuture.result
cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Index si_text is a duplicate of existing index test_v_nemesis"
```
[argus](https://argus.scylladb.com/test/0c756c99-fe0d-49aa-84bc-9c719bcafff8/runs?additionalRuns[]=a10402fd-91dd-4b27-b855-cda850aaa613)
test Id: a10402fd-91dd-4b27-b855-cda850aaa613

**Comment by fruch** [comment]
> @juliayakovlev , i encountered a related failure in a staging job:
> 
> ```
> 2023-08-31 09:15:37.584: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=53b45899-6b93-4012-9d33-e42316b7818d, source=LongevityTest.test_custom_time (longevity_test.LongevityTest)() message=Traceback (most recent call last):
> File "/home/ubuntu/scylla-cluster-tests/longevity_test.py", line 106, in test_custom_time
> self.run_prepare_write_cmd()
> File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 209, in run_prepare_write_cmd
> self.run_post_prepare_cql_cmds()
> File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 152, in run_post_prepare_cql_cmds
> self._run_cql_commands(post_prepare_cql_cmds)
> File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 143, in _run_cql_commands
> session.execute(cmd)
> File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/common.py", line 1751, in execute_verbose
> return execute_orig(*args, **kwargs)
> File "cassandra/cluster.py", line 2699, in cassandra.cluster.Session.execute
> File "cassandra/cluster.py", line 5018, in cassandra.cluster.ResponseFuture.result
> cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Index si_text is a duplicate of existing index test_v_nemesis"
> ```
> 
> [argus](https://argus.scylladb.com/test/0c756c99-fe0d-49aa-84bc-9c719bcafff8/runs?additionalRuns%5B%5D=a10402fd-91dd-4b27-b855-cda850aaa613) test Id: a10402fd-91dd-4b27-b855-cda850aaa613 #6565

please raise a new issue and assign it to @juliayakovlev

**Comment by juliayakovlev** [comment]
> > @juliayakovlev , i encountered a related failure in a staging job:
> > ```
> > 2023-08-31 09:15:37.584: (TestFrameworkEvent Severity.ERROR) period_type=one-time event_id=53b45899-6b93-4012-9d33-e42316b7818d, source=LongevityTest.test_custom_time (longevity_test.LongevityTest)() message=Traceback (most recent call last):
> > File "/home/ubuntu/scylla-cluster-tests/longevity_test.py", line 106, in test_custom_time
> > self.run_prepare_write_cmd()
> > File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 209, in run_prepare_write_cmd
> > self.run_post_prepare_cql_cmds()
> > File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 152, in run_post_prepare_cql_cmds
> > self._run_cql_commands(post_prepare_cql_cmds)
> > File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/loader_utils.py", line 143, in _run_cql_commands
> > session.execute(cmd)
> > File "/home/ubuntu/scylla-cluster-tests/sdcm/utils/common.py", line 1751, in execute_verbose
> > return execute_orig(*args, **kwargs)
> > File "cassandra/cluster.py", line 2699, in cassandra.cluster.Session.execute
> > File "cassandra/cluster.py", line 5018, in cassandra.cluster.ResponseFuture.result
> > cassandra.InvalidRequest: Error from server: code=2200 [Invalid query] message="Index si_text is a duplicate of existing index test_v_nemesis"
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > [argus](https://argus.scylladb.com/test/0c756c99-fe0d-49aa-84bc-9c719bcafff8/runs?additionalRuns%5B%5D=a10402fd-91dd-4b27-b855-cda850aaa613) test Id: a10402fd-91dd-4b27-b855-cda850aaa613 #6565
> 
> please raise a new issue and assign it to @juliayakovlev

@fruch 
Actually what I did in this PR - I fixed the index creation syntax error. This index was added in another commit with syntax mistake and was not created because of failure. And now when the mistake has been fixed and the index is started to be created + create_index nemesis - we got this error.

@yarongilor opened the issue

**Comment by juliayakovlev** [comment]
@fruch I did not find test that ran those nemeses after merge

**Comment by fruch** [comment]
> @fruch I did not find the tests that ran those nemeses after merge

It should run in large partition cases, right ?

we don't have a large partition case in tier1, but it isn't running enough time to reach that many nemesis.

@roydahan @fgelcer 
I wonder if we want to backport this new nemesis?

Especially that we don't yet have enough results of it, after it was merged...

**Comment by fgelcer** [comment]
> > @fruch I did not find the tests that ran those nemeses after merge
> 
> It should run in large partition cases, right ?
> 
> we don't have a large partition case in tier1, but it isn't running enough time to reach that many nemesis.
> 
> @roydahan @fgelcer I wonder if we want to backport this new nemesis?
> 
> Especially that we don't yet have enough results of it, after it was merged...

i would like to see that on 2023.1, but preferably after some soaking time in master ([we do have it on tier1](https://argus.scylladb.com/workspace?state=WyI5MTdlODI1Zi0xMWY5LTQ0OTMtYWNkYi1lYzUyNjZhM2FmNzgiXQ))

**Comment by juliayakovlev** [comment]
> > @fruch I did not find the tests that ran those nemeses after merge
> 
> It should run in large partition cases, right ?
> 
> we don't have a large partition case in tier1, but it isn't running enough time to reach that many nemesis.
> 
> @roydahan @fgelcer I wonder if we want to backport this new nemesis?
> 
> Especially that we don't yet have enough results of it, after it was merged...

Yes, it's for large partition tests. There is 200g 4 days test. But I did not find the delete nemeses in the list of the nemeses

**Comment by fruch** [comment]
> > > @fruch I did not find the tests that ran those nemeses after merge
> > 
> > 
> > It should run in large partition cases, right ?
> > we don't have a large partition case in tier1, but it isn't running enough time to reach that many nemesis.
> > @roydahan @fgelcer I wonder if we want to backport this new nemesis?
> > Especially that we don't yet have enough results of it, after it was merged...
> 
> i would like to see that on 2023.1, but preferably after some soaking time in master ([we do have it on tier1](https://argus.scylladb.com/workspace?state=WyI5MTdlODI1Zi0xMWY5LTQ0OTMtYWNkYi1lYzUyNjZhM2FmNzgiXQ))

it won't be socking, in such a short test case, i.e. we'll need to re-seed that case for that.

**Pull request merged**
